cmd/nodecmd/join.go

```diff
@@ -0,0 +1,479 @@
+	"github.com/ava-labs/avalanche-cli/pkg/constants"
+	"github.com/ava-labs/avalanche-cli/pkg/models"
+	"github.com/ava-labs/avalanche-cli/pkg/prompts"
+	"github.com/ava-labs/avalanche-cli/pkg/subnet"
+	"github.com/ava-labs/avalanche-cli/pkg/ux"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/spf13/cobra"
+)
+
+var (
+	deployTestnet   bool
+	deployMainnet   bool
+	keyName         string
+	subnetName      string
+	useLedger       bool
+	ledgerAddresses []string
+	weight          uint64
+	duration        time.Duration
+
+	ErrMutuallyExlusiveKeyLedger = errors.New("--key and --ledger,--ledger-addrs are mutually exclusive")
+	ErrStoredKeyOnMainnet        = errors.New("--key is not available for mainnet operations")
+	ErrNoBlockchainID            = errors.New("failed to find the blockchain ID for this subnet, has it been deployed/created on this network?")
+)
+
+func newJoinCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "join [clusterName]",
+		Short: "(ALPHA Warning) Join a subnet as a validator",
+		Long: `(ALPHA Warning) This command is currently in experimental mode.
+
+The node join command enables all nodes in a cluster to be validators of a Subnet.
+The command also enables nodes to be Primary Network validators by omitting the --subnet flag.
+If The command is run before the nodes are Primary Network validators, the command will first
+make the nodes Primary Network validators before making them Subnet validators. 
+If The command is run before the nodes are bootstrapped on the Primary Network, the command will fail. 
+You can check the bootstrap status by calling avalanche node status <clusteName>`,
+		SilenceUsage: true,
+		Args:         cobra.ExactArgs(1),
+		RunE:         joinSubnet,
+	}
+	cmd.Flags().StringVar(&subnetName, "subnet", "", "specify the subnet the node is validating")
+	cmd.Flags().BoolVarP(&deployTestnet, "testnet", "t", false, "set up validator in testnet (alias to `fuji`)")
+	cmd.Flags().BoolVarP(&deployTestnet, "fuji", "f", false, "set up validator in fuji (alias to `testnet`")
+	cmd.Flags().BoolVarP(&deployMainnet, "mainnet", "m", false, "set up validator in mainnet")
+	cmd.Flags().StringVarP(&keyName, "key", "k", "", "select the key to use [fuji only]")
+	cmd.Flags().StringSliceVar(&ledgerAddresses, "ledger-addrs", []string{}, "use the given ledger addresses")
+	cmd.Flags().Uint64Var(&weight, "stake-amount", 0, "how many AVAX to stake in the validator")
+	cmd.Flags().DurationVar(&duration, "staking-period", 0, "how long validator validates for after start time")
+
+	return cmd
+}
+
+func parseBootstrappedOutput(filePath string) (bool, error) {
+	jsonFile, err := os.Open(filePath)
+	if err != nil {
+		return false, err
+	}
+	defer jsonFile.Close()
+	byteValue, _ := io.ReadAll(jsonFile)
+	var result map[string]interface{}
+	if err := json.Unmarshal(byteValue, &result); err != nil {
+		return false, err
+	}
+	isBootstrappedInterface, ok := result["result"].(map[string]interface{})
```

Comment:
I think not ok cases should return error. same for other instances of this pattern